Allow users to review/delete comments, ratings and tags

rajaro opened the pull request:
Adds a new Comments and Ratings section to user account that allows users to see and delete their comments and ratings. 

TODO
- [x] Add Mink tests

rajaro committed:
Add Comments and Ratings section to account menu

rajaro committed:
Fix template indentation

rajaro committed:
Fix code style, add functions to service interfaces

demiankatz commented on languages/en.ini:
```diff
@@ -288,6 +290,7 @@ confirm_renew_all_text = "Do you wish to renew all your items?"
 confirm_renew_selected_text = "Do you wish to renew the selected items?"
 confirm_storage_retrieval_request_cancel_all_text = "Do you wish to cancel all your current storage retrieval requests?"
 confirm_storage_retrieval_request_cancel_selected_text = "Do you wish to cancel your selected storage retrieval requests?"
+confirm_usercomment_delete_selected_text = "Do you wish to delete your selected comments and ratings?"
```
We should probably think about how to label things so that the messages make sense even in systems where either comments or ratings are disabled, since it is possible to have one without the other. In this instance, replacing the phrase "comments and ratings" with "content" might do the job. For some of the labels, links and headings, it's possible we may need to introduce some logic to display the appropriate label, unless we can come up with a name for user-generated content that can apply to everything. (But since there are other things like tags that are not included, that might be tricky).

rajaro replied:
Could we use something like "your reviews", "user reviews" when dealing with comments and ratings?

demiankatz replied:
Yes, that's not a bad idea for a default. My only other idea would be something like "Record Feedback" which I think I like less. And, of course, it's easy enough to override the language strings locally if an institution wants to be more specific or use a different term. @sturkel89, do you have any opinions on this?

sturkel89 replied:
I don't it makes sense to use an entirely different term like **reviews** or **feedback** when the user is specifically interacting with features labeled **comments** and/or **ratings**. 

How about replacing **Comments and Ratings** with **Comments and/or Ratings**?

demiankatz replied:
The main problem I see with `Comments and/or Ratings` is that in a system where one of those two options does not exist, it's weird to mention a disabled feature. That's why I was looking for some broader terminology to cover user-generated content.

rajaro replied:
Now with the tabbed solution, each tab has their own (comments, ratings, tags) texts, but for the account menu button I went with Reviews. Not sure if it quite describes tags though. Should we think of some other term now that tags are included as well?

demiankatz replied:
The best I can think of is "Content" or "User Content." If we want a verb, maybe the menu item could be something like "Manage User Content" or "Manage Submitted Content." I don't really love this, so I'm open to better ideas if somebody has any!

demiankatz commented on module/VuFind/src/VuFind/Config/AccountCapabilities.php:
```diff
@@ -89,6 +89,16 @@ public function getCommentSetting()
+     */
+    public function getRatingSetting(): bool
+    {
+        return !empty($this->config->Social->rating);
```
Might it make sense to return enabled / disabled strings here similar to getCommentSetting, just in case we want to add additional rating modes in the future? Or are we confident enough that additional nuance will not be added that we should stick with the simpler boolean?

demiankatz commented on module/VuFind/src/VuFind/Controller/CommentsController.php:
```diff
@@ -0,0 +1,123 @@
+        ) {
+            throw new ForbiddenException('Comments and ratings disabled.');
+        }
+        $limit = 50;
```
Should we make this configurable?

demiankatz commented on module/VuFind/src/VuFind/Controller/CommentsController.php:
```diff
@@ -0,0 +1,123 @@
+        }
+
+        return $this->redirect()->toRoute(
+            'default',
```
I think it would be simpler to redirect to the `comments-usercomments` route here instead of using `default` with parameters.

demiankatz commented on module/VuFind/src/VuFind/Db/Table/Comments.php:
```diff
@@ -174,4 +174,89 @@ public function getStatistics()
+        } elseif ($comments) {
+            $paginatedSelect = $commentSelect;
+        }
+        $paginatedSelect->order('created DESC');
```
Would it make sense to add a sort parameter here, so we could potentially sort by created ascending/created descending/title?

rajaro replied:
Yes. In addition to sorting, should we add the option to filter either comments or ratings?

demiankatz replied:
I suppose you could change the `$comments` and `$ratings` arguments to accept something more nuanced than a boolean, and then use them for filtering -- so "true" for all values, or a filter. But I'm not sure what filter would make the most sense here. I'm certainly open to the idea, but we shouldn't make this more complicated than we need it to be. :-)

sturkel89 replied:
I like the idea of allowing the user to sort by title and by created ascending/descending. 

It looks like the default sort is by date/time created, with the most recent comment or rating listed first.

demiankatz commented on module/VuFind/src/VuFind/Navigation/AccountMenu.php:
```diff
@@ -172,6 +172,13 @@ public static function getDefaultMenuConfig(): array
                         'checkMethod' => 'checkHistory',
                     ],
+                    [
+                        'name' => 'usercomments',
```
Equivalent configuration should also be added to [AccountMenu.yaml](https://github.com/vufind-org/vufind/blob/dev/config/vufind/AccountMenu.yaml).

demiankatz commented on themes/bootstrap5/templates/comments/usercomments.phtml:
```diff
@@ -0,0 +1,94 @@
+              </td>
+              <td class="user-comment">
+                <?php if (isset($comment['rating'])): ?>
+                  <?=$this->render('Helpers/star-rating.phtml', ['readonly' => true, 'ratingData' => ['rating' => $comment['rating'], 'count' => 1]])?>
```
This will require minor adjustment if/when #4317 is merged -- just something to keep an eye on depending on which part gets done first.

demiankatz reviewed (changes requested):
Thanks, @rajaro, this is a great idea. I haven't tried hands-on testing yet, but I did a quick readthrough of the diffs here. See below for some comments and suggestions.

rajaro committed:
requested changes (accountmenu.yaml, configuration settings handling, some smaller fixes)

demiankatz commented on module/VuFind/src/VuFind/Controller/AbstractBase.php:
```diff
@@ -655,6 +655,17 @@ protected function commentsEnabled()
+     *
+     * @return bool
+     */
+    protected function ratingsEnabled()
```
Probably worth putting a return type on this (I try to type anything new that I add):

```suggestion
    protected function ratingsEnabled(): bool
```

demiankatz commented on module/VuFind/src/VuFind/Controller/CommentsController.php:
```diff
@@ -0,0 +1,122 @@
+        $sort = $this->params()->fromQuery('sort', '');
+        $service = $this->getDbService(\VuFind\Db\Service\CommentsServiceInterface::class);
+        $comments = $service->getCommentsAndRatingsByUserId(
+            $user->id,
```
Direct property access on the user object is deprecated -- we should use the UserEntityInterface getter instead:

```suggestion
            $user->getId(),
```

demiankatz commented on module/VuFind/src/VuFind/Db/Service/CommentsService.php:
```diff
@@ -165,4 +165,51 @@ public function changeResourceId(int $old, int $new): void
+        string $sort,
+        bool $comments,
+        bool $ratings
+    ): \Laminas\Paginator\Paginator {
```
The CommentsServiceInterface has a return type of `Paginator|array` and this implementation only returns `Paginator`. I'd suggest that we should pick one or the other:

Option 1: Let's just return an array; it's simpler and more future-proof.

Option 2: If we really need a Paginator, let's make that explicit in the interface and also rename the method to include the word "Paginator" (like other database server getters that return paginators).

Generally speaking, I prefer option 1 if it's practical, but I'm okay with option 2 if it's easier.

demiankatz commented on module/VuFind/src/VuFind/Navigation/AccountMenu.php:
```diff
@@ -352,6 +359,17 @@ protected function checkIlsFunction(string $function): bool
+    protected function checkCommentsAndRatings(): bool
+    {
+        return ('enabled' === $this->accountCapabilities->getCommentSetting())
+            || ('enabked' === $this->accountCapabilities->getRatingSetting());
```
Typo:

```suggestion
            || ('enabled' === $this->accountCapabilities->getRatingSetting());
```

demiankatz reviewed (changes requested):
Thanks for the progress here, @rajaro! See below for a few new comments from another look through the code.

rajaro committed:
Use paginator as the only return value and also in function names. Other smaller requested changes

sturkel89 commented:
I tested this feature in its current state, adding both comments AND ratings to two items, and just ratings or just comments to two others. See this screenshot:

![image](https://github.com/user-attachments/assets/dca426e7-4970-4949-9196-622814eb9e4d)

I think it would be better if all comments and ratings for the same record were grouped together in this interface in a single row/entry. They're grouped together in the Comments tab on the record itself already:

![image](https://github.com/user-attachments/assets/317c5027-a073-4101-a364-a95bb48a7d00)

-------------

I also suggest adding instructions for editing comments and ratings at the top of the **Your Comments and/or Ratings** page in the Your Account section. The instructions could say "To edit comments or ratings, click the item's Title and make your edits in the record."

rajaro commented:
@sturkel89 If comments and ratings are grouped together in a single row, how should deletion be handled? Users can have multiple comments but only one rating for a record and the rating would be displayed next to all the comments for the record. Should we then have separate buttons like "Delete selected ratings" and "Delete selected comments" so the user can delete a single comment without deleting the rating and vice versa?

demiankatz commented:
@rajaro, I guess to some extent, interleaving comments and ratings in the same list is confusing no matter how you do it. Is it worth considering handling them separately, either through a tabbed interface or two separate lists on the page? I'm not saying we have to change the approach, just trying to brainstorm all the options. There are some obvious advantages to keeping everything grouped together, but on the other hand, advantages to keeping things separate include simplifying our labeling problem and reducing the complexity of our database calls (which will have to be translated into Doctrine very soon for compatibility with the coming #2233 -- a somewhat daunting task for the current logic).

demiankatz commented:
Also, a separate thought: is there a reason we don't also include tags in this mechanism? As noted in [VUFIND-905](https://openlibraryfoundation.atlassian.net/issues/VUFIND-905), there is currently a gap in tag management functionality. If we're trying to create comprehensive visibility/control for user-generated data, leaving tags out seems like a significant omissions. Even if we don't want to address it right away, we should keep it in mind during our design process to ensure that it can be logically added in the future.

sturkel89 commented:
@rajaro - To answer your question, I agree that separate buttons for  "Delete selected ratings" and "Delete selected comments" would make sense.

But I think that Demian's suggestions of multiple lists, or multiple tabs (and hopefully a tab for TAGS as well -- so three tabs in total) would be the optimal solution.

rajaro commented:
The tab solution does sound good, as it would simplify things quite a bit. And especially if we want to include tags in this, something like a tabbed interface seems like the way to go

rajaro committed:
Use tabbed interface. List user tags as well. Add sort menu.

rajaro committed:
phpcs fixes

rajaro committed:
Use User Content instead of User Reviews. Fix sort menu positioning

demiankatz commented on themes/bootstrap5/theme.config.php:
```diff
@@ -170,6 +170,7 @@
             'collapse' => 'Collapse:_', // uses the icons below
             'collapse-close' => 'FontAwesome:chevron-up',
             'collapse-open' => 'FontAwesome:chevron-down',
+            'comment' => 'FontAwesome:comment',
```
Should we rename this icon to `user-content`?

demiankatz commented on config/vufind/AccountMenu.yaml:
```diff
@@ -97,6 +97,12 @@ Account:
+    - name: usercontent
+      label: user_content
+      route: myresearch-usercontent
+      icon: comment
```
Should we rename this (see previous comment)?

demiankatz commented on module/VuFind/src/VuFind/Controller/CommentsController.php:
```diff
@@ -0,0 +1,133 @@
+     *
+     * @return View
+     */
+    public function userCommentsAction()
```
I wonder if there's a better, less redundant action name -- one that would result in a cleaner URL and require less complicated logic for building the route name for redirection. Maybe something like userListAction or userOwnedAction or userCreatedAction?

demiankatz commented on module/VuFind/src/VuFind/Controller/RatingsController.php:
```diff
@@ -0,0 +1,134 @@
+            $page,
+            $sort,
+        );
+        $sortList = [
```
Maybe it's worth programmatically generating this list from an array of `['created desc', 'created asc', 'title']` to reduce redundancy and make future customization/expansion easier.

demiankatz commented on module/VuFind/src/VuFind/Controller/RatingsController.php:
```diff
@@ -0,0 +1,134 @@
+        }
+        $records = $recordLoader->loadBatch($ids, true);
+        foreach ($ratings as $i => $c) {
+            $c['recordTitle'] = $records[$i]->getTitle() ?? '';
```
Does this do anything? I thought `$c` would have to be a reference in order for changes to get applied to it. Maybe I'm misunderstanding/misremembering how arrays are handled in foreach loops, though!

rajaro replied:
`$c` here is actually an ArrayObject, so it works as it is passed as a reference by default. This part is also now moved to UserContentTrait

demiankatz replied:
I don't think we can rely on these being ArrayObjects in the future -- that is a quirk of Laminas\Db, but we are in the process of migrating to Doctrine (see #2233) which will I believe will use plain arrays in this situation. (If I'm remembering correctly, it will return entity objects when selecting from a single table, and arrays when selecting from joined tables). Maybe that's just a problem we need to figure out when merging this work into #2233, but anything we can do to use "less exotic" return values out of database services may make that work less complicated when we get there.

rajaro replied:
Now it is `&$c` which should keep this working with plain arrays

demiankatz commented on module/VuFind/src/VuFind/Controller/TagController.php:
```diff
@@ -66,4 +67,94 @@ public function homeAction()
+            $page,
+            $limit,
+        );
+        $sortList = [
```
Same comment as above about potential programmatic generation...

demiankatz commented on module/VuFind/src/VuFind/Controller/CommentsController.php:
```diff
@@ -0,0 +1,133 @@
+            $sort,
+        );
+
+        $sortList = [
```
Another candidate for programmatic generation. Maybe we could even have a trait for processing these lists.

demiankatz reviewed (changes requested):
I ran out of time before I could do a full review of everything here, but see below for a few suggestions I was able to make during a partial review!

I've also asked @sturkel89 to take another look and comment on the revised functionality.

rajaro committed:
Add UserContentTrait to make code cleaner. Some renaming. Fix typo

demiankatz commented on module/VuFind/src/VuFind/Controller/UserContentTrait.php:
```diff
@@ -0,0 +1,84 @@
+ * @link     https://vufind.org Main Site
+ */
+
+namespace VuFind\Controller;
```
Do you mind moving this into the `VuFind\Controller\Feature` namespace? We haven't been consistent in the past, but going forward we're trying to keep all of our traits in "Feature" namespaces.

demiankatz reviewed (changes requested):
See below for one small request, in addition to the ongoing discussion about paginator return values from database services.

rajaro committed:
Move UserContentTrait to Feature namespace. Use reference when adding record titles

demiankatz commented on module/VuFind/config/module.config.php:
```diff
@@ -799,6 +805,7 @@
     'Cart/Processor', 'Cart/Save', 'Cart/SearchResultsBulk',
     'Channels/Home', 'Channels/Record', 'Channels/Search',
     'Checkouts/History', 'Checkouts/PurgeHistory',
+    'Comments/UserList', 'Comments/DeleteComments',
```
I'd reverse these for the sake of alphabetical order:

```suggestion
    'Comments/DeleteComments', 'Comments/UserList',
```

demiankatz commented on module/VuFind/config/module.config.php:
```diff
@@ -827,13 +834,14 @@
     'Primo/Advanced', 'Primo/CitedBy', 'Primo/Cites', 'Primo/Home', 'Primo/Search',
     'ProQuestFSG/Advanced', 'ProQuestFSG/Home', 'ProQuestFSG/Results',
     'QRCode/Show', 'QRCode/Unavailable', 'Records/Home',
+    'Ratings/UserList', 'Ratings/DeleteRatings',
```
Same here:

```suggestion
    'Ratings/DeleteRatings', 'Ratings/UserList',
```

demiankatz commented on module/VuFind/src/VuFind/Controller/Feature/UserContentTrait.php:
```diff
@@ -0,0 +1,84 @@
+    public function getUserContentRecordTitles(\Laminas\Paginator\Paginator $contents)
+    {
+        $recordLoader = $this->serviceLocator->get(\VuFind\Record\Loader::class);
+        $ids = [];
+        foreach ($contents as $content) {
+            $ids[] = $content['source'] . '|' . $content['record_id'];
+        }
```
Entirely optional, but I think you might be able to optimize this to something like the below, if I'm remembering my arrow function syntax correctly:

```suggestion
        $ids = array_map(fn ($content) => $content['source'] . '|' . $content['record_id'], $contents);
```

rajaro replied:
Doesn't seem to work as `array_map` expects an array but `$contents` is a Paginator

demiankatz replied:
You could use `iterator_to_array` to make `$contents` into an array, and then it will probably work. Just a question of whether the functional approach is preferable or not -- I'll leave it up to you! :-)

rajaro replied:
Switched to arrow function

demiankatz commented on module/VuFind/src/VuFind/Db/Service/CommentsServiceInterface.php:
```diff
@@ -95,6 +95,16 @@ public function deleteIfOwnedByUser(int $id, UserEntityInterface|int $userOrId):
     public function deleteByUser(UserEntityInterface|int $userOrId): void;
 
+    /**
+     * Deletes comments by given comment ids and user id
```
Add punctuation for consistency.

```suggestion
     * Deletes comments by given comment ids and user id.
```

demiankatz commented on module/VuFind/src/VuFind/Db/Service/CommentsServiceInterface.php:
```diff
@@ -111,6 +121,23 @@ public function getStatistics(): array;
     public function getCommentById(int $id): ?CommentsEntityInterface;
 
+    /**
+     * Get a paginated result of all comments by user id
```
Add punctuation for consistency.

```suggestion
     * Get a paginated result of all comments by user id.
```

demiankatz commented on module/VuFind/src/VuFind/Db/Service/RatingsServiceInterface.php:
```diff
@@ -101,4 +111,21 @@ public function addOrUpdateRating(
     ): int;
+
+    /**
+     * Get a paginated result of all ratings by user id
```
Add punctuation for consistency.

```suggestion
     * Get a paginated result of all ratings by user id.
```

demiankatz commented on module/VuFind/src/VuFind/Db/Service/RatingsServiceInterface.php:
```diff
@@ -79,6 +79,16 @@ public function getCountsForRecord(
     public function deleteByUser(UserEntityInterface|int $userOrId): void;
 
+    /**
+     * Deletes ratings by given rating ids and user id
```
Add punctuation for consistency.

```suggestion
     * Deletes ratings by given rating ids and user id.
```

demiankatz commented on module/VuFind/src/VuFind/Db/Service/RatingsService.php:
```diff
@@ -126,4 +126,45 @@ public function addOrUpdateRating(
     }
+
+    /**
+     * Delete ratings by given user and rating ids
```
Add punctuation for consistency.

```suggestion
     * Delete ratings by given user and rating ids.
```

demiankatz commented on module/VuFind/src/VuFind/Db/Service/RatingsService.php:
```diff
@@ -126,4 +126,45 @@ public function addOrUpdateRating(
+    }
+
+    /**
+     * Get a paginated result of all ratings by user id
```
Add punctuation for consistency.

```suggestion
     * Get a paginated result of all ratings by user id.
```

demiankatz commented on module/VuFind/src/VuFind/Db/Table/Comments.php:
```diff
@@ -174,4 +174,55 @@ public function getStatistics()
+
+    /**
+     * Get a paginated result of all comments
+     * made by the user
```
Add punctuation (and collapse to one line) for consistency.

```suggestion
     * Get a paginated result of all comments made by the user.
```

demiankatz commented on module/VuFind/src/VuFind/Db/Table/Comments.php:
```diff
@@ -174,4 +174,55 @@ public function getStatistics()
+        if ($page > 0) {
+            $commentSelect->offset($page);
+        }
+        if (null !== $limit) {
```
Since `$limit` is typed to `int`, I don't think you need this null check; shouldn't be possible for the value to be null.

demiankatz commented on module/VuFind/src/VuFind/Db/Table/Comments.php:
```diff
@@ -174,4 +174,55 @@ public function getStatistics()
+        $adapter = new \Laminas\Paginator\Adapter\LaminasDb\DbSelect($commentSelect, $this->getSql());
+        $paginator = new \Laminas\Paginator\Paginator($adapter);
+        $paginator->setItemCountPerPage($limit);
+        if (null !== $page) {
```
This value also cannot be null.

demiankatz commented on module/VuFind/src/VuFind/Db/Table/Ratings.php:
```diff
@@ -234,4 +234,55 @@ public function getStatistics(): array
     }
+
+    /**
+     * Get a paginated result of all ratings
+     * made by the user
```
More minor comment standardization.

```suggestion
     * Get a paginated result of all ratings made by the user.
```

demiankatz commented on module/VuFind/src/VuFind/Db/Table/Ratings.php:
```diff
@@ -234,4 +234,55 @@ public function getStatistics(): array
+        if ($page > 0) {
+            $ratingSelect->offset($page);
+        }
+        if (null !== $limit) {
```
Unnecessary null check.

demiankatz commented on module/VuFind/src/VuFind/Db/Table/Ratings.php:
```diff
@@ -234,4 +234,55 @@ public function getStatistics(): array
+        $adapter = new \Laminas\Paginator\Adapter\LaminasDb\DbSelect($ratingSelect, $this->getSql());
+        $paginator = new \Laminas\Paginator\Paginator($adapter);
+        $paginator->setItemCountPerPage($limit);
+        if (null !== $page) {
```
Unnecessary null check.

demiankatz commented on module/VuFind/src/VuFind/Controller/CommentsController.php:
```diff
@@ -0,0 +1,121 @@
+        }
+        $limit = $this->getService(\VuFind\Config\AccountCapabilities::class)->getUserContentPageSize();
+        $page = $this->params()->fromQuery('page', 1);
+        $sort = $this->params()->fromQuery('sort', 'created desc');
```
We should probably do some sort of validation on sort, here and/or in the service itself. (I think it makes the most sense to validate here, so that if we add the ability to disable sort options in the future, the limit cannot be circumvented by modifying the URL directly).

demiankatz replied:
Might also make sense to validate that $page is a positive integer as well.

rajaro replied:
Some validation added and also moved these to UserContentTrait

demiankatz commented on module/VuFind/src/VuFind/Controller/RatingsController.php:
```diff
@@ -0,0 +1,123 @@
+        }
+        $limit = $this->getService(\VuFind\Config\AccountCapabilities::class)->getUserContentPageSize();
+        $page = $this->params()->fromQuery('page', 1);
+        $sort = $this->params()->fromQuery('sort', 'created desc');
```
Same validation comments apply here.

demiankatz commented on module/VuFind/src/VuFind/Controller/TagController.php:
```diff
@@ -66,4 +80,68 @@ public function homeAction()
+        }
+        $limit = $this->getService(\VuFind\Config\AccountCapabilities::class)->getUserContentPageSize();
+        $page = $this->params()->fromQuery('page', 1);
+        $sort = $this->params()->fromQuery('sort', 'posted desc');
```
...and here as well.

demiankatz commented on module/VuFind/src/VuFind/Db/Service/CommentsService.php:
```diff
@@ -165,4 +165,45 @@ public function changeResourceId(int $old, int $new): void
     }
+
+    /**
+     * Get a paginated result of all comments by user id
```
Add punctuation for consistency.

```suggestion
     * Get a paginated result of all comments by user id.
```

demiankatz commented on module/VuFind/src/VuFind/Db/Service/CommentsService.php:
```diff
@@ -165,4 +165,45 @@ public function changeResourceId(int $old, int $new): void
+    }
+
+    /**
+     * Delete comments by given user and comment ids
```
Add punctuation for consistency.

```suggestion
     * Delete comments by given user and comment ids.
```

demiankatz commented on themes/bootstrap5/templates/comments/userlist.phtml:
```diff
@@ -0,0 +1,104 @@
+<div class="<?=$this->layoutClass('mainbody')?>">
+  <ul class="nav nav-tabs" role="tablist">
+    <?php foreach ($tabs as $tab): ?>
+      <li class="nav-item" role="presentation" id="user-content-<?=$tab?>" data-tab="<?=$this->escapeHtmlAttr($tab)?>"><a class="<?='comments' === $tab ? 'active' : ''?> nav-link" href="<?=$this->url($tab . '-' . 'userlist')?>"><?=$this->transEsc(ucfirst($tab))?></a></li>
```
Is there a reason that you escape `$tab` in `data-tab` but not in `id`? I think it would make sense to be consistent.

demiankatz commented on themes/bootstrap5/templates/ratings/userlist.phtml:
```diff
@@ -0,0 +1,102 @@
+<div class="<?=$this->layoutClass('mainbody')?>">
+  <ul class="nav nav-tabs" role="tablist">
+    <?php foreach ($tabs as $tab): ?>
+      <li class="nav-item" role="presentation" id="user-content-<?=$tab?>" data-tab="<?=$this->escapeHtmlAttr($tab)?>"><a class="<?='ratings' === $tab ? 'active' : ''?> nav-link" href="<?=$this->url($tab . '-' . 'userlist')?>"><?=$this->transEsc(ucfirst($tab))?></a></li>
```
Is there a reason that you escape `$tab` in `data-tab` but not in `id`? I think it would make sense to be consistent.

demiankatz commented on themes/bootstrap5/templates/tag/userlist.phtml:
```diff
@@ -0,0 +1,97 @@
+<div class="<?=$this->layoutClass('mainbody')?>">
+  <ul class="nav nav-tabs" role="tablist">
+    <?php foreach ($tabs as $tab): ?>
+      <li class="nav-item" role="presentation" id="user-content-<?=$tab?>" data-tab="<?=$this->escapeHtmlAttr($tab)?>"><a class="<?='tag' === $tab ? 'active' : ''?> nav-link" href="<?=$this->url($tab . '-' . 'userlist')?>"><?=$this->transEsc(ucfirst($tab))?></a></li>
```
Is there a reason that you escape `$tab` in `data-tab` but not in `id`? I think it would make sense to be consistent.

demiankatz commented on themes/bootstrap5/templates/comments/userlist.phtml:
```diff
@@ -0,0 +1,104 @@
+                  <div class="checkbox">
+                    <label>
+                      <span class="sr-only"><?=$this->transEsc('Select');?></span>
+                      <input class="checkbox-select-item" type="checkbox" name="deleteSelectedcomment[]" value="<?=$this->escapeHtmlAttr(($comment['id'] ?? ''))?>" id="checkbox_comment_<?=$comment['id']?>">
```
Inconsistent handling of `$comment['id']` here -- if we need null coalescing, maybe we should assign the value in a separate line so we don't have to do it repeatedly.

demiankatz commented on themes/bootstrap5/templates/comments/userlist.phtml:
```diff
@@ -0,0 +1,104 @@
+              </td>
+              <td data-label="<?=$this->transEscAttr('Created')?>"><?=$this->dateTime()->convertToDisplayDateAndTime('Y-m-d H:i:s', $comment['created'])?></td>
+              <td data-label="<?=$this->transEscAttr('Title')?>">
+                <a href="<?=$this->url('record', ['id' => $comment['record_id']])?>"><?=$comment['recordTitle']?></a>
```
`$comment['recordTitle']` should probably be escaped.

demiankatz commented on themes/bootstrap5/templates/ratings/userlist.phtml:
```diff
@@ -0,0 +1,102 @@
+                    <div class="checkbox">
+                      <label>
+                        <span class="sr-only"><?=$this->transEsc('Select');?></span>
+                        <input class="checkbox-select-item" type="checkbox" name="deleteSelectedrating[]" value="<?=$this->escapeHtmlAttr(($rating['id'] ?? ''))?>" id="checkbox_rating_<?=$rating['id']?>">
```
Inconsistent handling of `$rating['id']`.

demiankatz commented on themes/bootstrap5/templates/ratings/userlist.phtml:
```diff
@@ -0,0 +1,102 @@
+              </td>
+              <td data-label="<?=$this->transEscAttr('Created')?>"><?=$this->dateTime()->convertToDisplayDateAndTime('Y-m-d H:i:s', $rating['created'])?></td>
+              <td data-label="<?=$this->transEscAttr('Title')?>">
+                <a href="<?=$this->url('record', ['id' => $rating['record_id']])?>"><?=$rating['recordTitle']?></a>
```
Missing escaping of `$rating['recordTitle']`.

demiankatz commented on themes/bootstrap5/templates/tag/userlist.phtml:
```diff
@@ -0,0 +1,97 @@
+                  <div class="checkbox">
+                    <label>
+                    <span class="sr-only"><?=$this->transEsc('Select');?></span>
+                    <input class="checkbox-select-item" type="checkbox" name="deleteSelectedtag[]" value="<?=$this->escapeHtmlAttr(($tag['id'] ?? ''))?>" id="checkbox_tag_<?=$tag['id']?>">
```
Inconsistent handling of `$tag['id']`.

demiankatz commented on themes/bootstrap5/templates/tag/userlist.phtml:
```diff
@@ -0,0 +1,97 @@
+              </td>
+              <td data-label="<?=$this->transEscAttr('Created')?>"><?=$this->dateTime()->convertToDisplayDateAndTime('Y-m-d H:i:s', $tag['posted'])?></td>
+              <td data-label="<?=$this->transEscAttr('Title')?>">
+                <a href="<?=$this->url('record', ['id' => $tag['record_id']])?>"><?=$tag['recordTitle']?></a>
```
Missing escaping of `$tag['recordTitle']`.

demiankatz commented on module/VuFind/src/VuFind/Db/Table/ResourceTags.php:
```diff
@@ -706,7 +706,7 @@ public function getResourceTags(
             ['r' => 'resource'],
             'resource_tags.resource_id = r.id',
-            ['title' => 'title']
+            ['title' => 'title', 'record_id', 'source']
```
Do we really need the `'title' => 'title'` syntax here, or can we simplify to just `'title'`?

demiankatz reviewed (changes requested):
Thanks, @rajaro! I've given this a more thorough review today. Lots of comments, but most are very minor (and I tried to make it easy for you to just apply my suggestions to all of the comment-related nitpicks... but please note that my suggested arrow function code simplification is totally untested and might be incorrect!).

rajaro committed:
Update module/VuFind/config/module.config.php

Co-authored-by: Demian Katz <demian.katz@villanova.edu>

rajaro committed:
Update module/VuFind/src/VuFind/Db/Service/CommentsServiceInterface.php

Co-authored-by: Demian Katz <demian.katz@villanova.edu>

rajaro committed:
Update module/VuFind/src/VuFind/Db/Service/CommentsServiceInterface.php

Co-authored-by: Demian Katz <demian.katz@villanova.edu>

rajaro committed:
Update module/VuFind/config/module.config.php

Co-authored-by: Demian Katz <demian.katz@villanova.edu>

rajaro committed:
Add missing escapes and other requested changes

rajaro committed:
Merge remote-tracking branch 'origin/user-comments-ratings' into user-comments-ratings

rajaro committed:
Add validation for sort and page parameters. More smaller requested changes

rajaro committed:
Add missing escape

demiankatz commented:
Thanks, @rajaro, it looks like you have addressed all of my feedback (apart from the functional programming optimization that is still under discussion, but which is not necessary to move forward).

I think now this is ready for @sturkel89 to do another round of hands-on testing; depending on what she finds there, we can decide how to proceed.

There is one other important thing that we should do before merging this: add a Mink test case to cover all of the new functionality. Because of #2233, we will need to rewrite all of the low-level database code soon, so it is important to have robust test cases that ensure that nothing breaks when we transition to Doctrine. Please let me know if you need help from my end on that!

sturkel89 reviewed (changes requested):
This feature is looking great! The switch to a tabbed interface is a big improvement. The sorting works well. Great work! This is a very nice new feature. I have a few tiny suggestions for tweaks; see below.

I agree with Demian above that the label "Manage User Content" is accurate but somehow doesn't feel like the best solution. I would like to propose something different, but haven't thought of anything yet.

-------------

### Label for Tag - FIXED!
The terms **Comments** and **Ratings** are both plural, so I think the label **Tag** should be changed to **Tags**.

![image](https://github.com/user-attachments/assets/580b8362-57fc-4f55-8750-1a5d4af736e5)

---

### Checkboxes: Formatting and clickability - ALL FIXED except floating line
In each area, there seems to be an extra floating line near the checkboxes, and each row is a big taller (more padding) than seems necessary. Can you reformat this slightly?

![image](https://github.com/user-attachments/assets/714a5474-ebba-449c-8ca8-b507e5f09d86)

Also, I see that the clickable area is very tight around the checkboxes; the user has to have the mouse right on the box or its borders. The clickable area around the checkboxes in other areas is slightly bigger which is helpful for accessibility. If you can see how the checkboxes are set up in Saved Items, you'll see what I mean. Can you enlarge the clickable area around checkboxes in this feature?

---

### Instructions for Ratings - FIXED!
Ratings is the only kind of user content that can be changed, not just added and deleted. You have helpfully included instructions:

<img src="https://github.com/user-attachments/assets/d824b900-3001-4be0-8537-c4c34f58d467" width = 600>


I think the instructions would fit better ABOVE the "Showing 1-3 of 3 items line, and would benefit from a bit more padding setting them off. Could the text say:

**Ratings can be changed. To edit ratings, click the item's title and make your edits in the record.**

Also, I notice that the ratings stars appear to be clickable in this view, as the mouse turns to a pointing hand - but it isn't actually possible to click them. Can you adjust this so the area containing the stars does not get a pointing hand?

<img src="https://github.com/user-attachments/assets/f2459616-be9e-4849-9787-4ea72e9c2de6" width = 300>

demiankatz commented:
Regarding the ratings hand, is there a reason why we can't actually make the ratings editable in this view? If we could do that, the instructions would be unnecessary and users wouldn't have to go through as many steps. Maybe there's a reason that we can't, but it seems like it should be possible in theory!

rajaro committed:
Allow editing ratings from ratings-userlist. Requested style changes and fixes

rajaro commented:
@demiankatz Ratings are now editable in this view. And I'll start to work on Mink tests. 

@sturkel89 All those should now be done. Except for the extra floating line near the checkboxes, I'm not seeing it locally but I'll look into it

demiankatz commented on themes/bootstrap5/templates/ratings/userlist.phtml:
```diff
@@ -0,0 +1,103 @@
+              </td>
+              <td class="user-rating" data-label="<?=$this->transEscAttr('Rating')?>">
+                <a href="<?=$this->escapeHtmlAttr($this->recordLinker()->getActionUrl($rating['source'] . '|' . $rating['record_id'], 'Rating'))?>" data-lightbox>
+                  <?=$this->render('Helpers/star-rating.phtml', ['readonly' => true, 'ratingData' => ['rating' => $rating['rating'], 'count' => 1]])?>
```
When you merge the latest dev code into your branch, you'll need to change this to:

```suggestion
                  <?=$this->component('star-rating', ['readonly' => true, 'ratingData' => ['rating' => $rating['rating'], 'count' => 1]])?>
```

...because of #4317 being completed.

rajaro replied:
Done

demiankatz reviewed (changes requested):
Thanks, @rajaro, your improvements all look very good. One of my earlier comments has just become relevant; you should merge the latest dev branch and make a minor adjustment. See below for more details.

sturkel89 reviewed (changes requested):
This is all looking great to me! As you say, the only remaining issue from my earlier review is the horizontal line that's a few spaces below each checkbox.

Here is some information from the CSS Inspector; I think it's the **border bottom** of the **checkbox-container result**, or something like that.

![image](https://github.com/user-attachments/assets/9e3851ac-f053-47b1-8e70-57ad41ee540d)

--------------

### Column width in Tags tab

Upon further testing, I notice that the User Content area has some formatting oddities if the user chooses to add comments or tags to an item with a VERY long title. See below for examples with short tags, vs. examples with multi-word comments. I think you could probably tweak the tags and comments settings a tiny bit to make things even better.

### RATINGS
This looks fine to me.

![image](https://github.com/user-attachments/assets/663a3583-6cb2-4c35-b4a4-6a739c0407e4)

### COMMENTS
It's interesting that the 'More' and 'Less' expanders appear for multi-line comments, regardless of the space available for them. This screenshots shows both comments opened; originally only the first line was visible, with the More expander. I don't think the expanders are needed; why not display the entire comment?

![image](https://github.com/user-attachments/assets/152c993e-e7a0-4395-912d-125076e5d67f)

These expanders don't appear with shorter book titles:

![image](https://github.com/user-attachments/assets/de777ba8-843f-4fa6-bddd-72a5240fb3f3)


### TAGS 

The Tags column is three characters wide which leads to weird word wrapping.

![image](https://github.com/user-attachments/assets/8bf86da4-d3c6-43c7-abf6-96746cfe946f)

The tags look fine with shorter book titles:

![image](https://github.com/user-attachments/assets/3dbb4924-c038-4534-81d8-ad8189eb4a5a)

rajaro committed:
Requested style changes

rajaro committed:
Merge remote-tracking branch 'vufind/dev' into user-comments-ratings

rajaro committed:
Use star rating component

rajaro commented:
@sturkel89 The extra line near checkboxes should now be fixed. I was testing with bootstrap5 theme instead of sandal5 which is why I didn't see the line at first. With regards to the 'more' and 'less' functions for comments, I'm using a truncate helper there that requires a defined amount of rows to display before truncating. Now I changed the value to 5, so only after 5 rows of comments would the 'more' button be displayed, so it only applies to longer comments. 

Also the comments column should already have a set width (that I now also added for Tags), so the length of the title should no longer change the way the comment/tag is positioned

demiankatz commented:
Thanks for the progress, @rajaro! I'm going to be at a conference for the rest of this week, but I imagine @sturkel89 will have time to validate the rest of your fixes while I'm away, and then I can hopefully complete the merge next week if no major problems arise.

sturkel89 reviewed (changes requested):
Thanks for making some improvements! 

- I see the difference from your changes to the truncate helper setting; now my full comments appear without More or Less expanders.

- I see that the extra line doesn't appear in bootstrap5. 

- **However, I still see the line in sandal5.**

- **Also, I still see the Tags column as being very narrow when there's a long title on the list, in both themes.**

See below for screenshots. Thanks!

---

### Sandal5 - comments look good! Line is present.

![image](https://github.com/user-attachments/assets/dd2d8c9b-d9ed-4ec4-b7e7-d3a0c71bdf8d)

---

### Sandal5 - line is present and tags column is narrow

![image](https://github.com/user-attachments/assets/1c038338-0d63-4ab0-b096-4fa7380150ae)

---

### Bootstrap5 - line is absent, but tags column is narrow

![image](https://github.com/user-attachments/assets/9de503e9-aafa-4fbd-b423-d178910cd3c1)

rajaro commented:
@sturkel89 That's odd. It seems to me like the new SCSS rules added in this PR are not applied on your end. If I'm not mistaken, you should make sure to build the scss to have the new rules apply.

demiankatz commented:
@rajaro, we always commit the compiled.css file; is it possible that you didn't do this after your last updates? That might explain everything.

(There has been some discussion about adjusting this workflow in future, but for now we're still committing these artifacts).

rajaro committed:
Add compiled.css

sturkel89 commented:
At @demiankatz's suggestion, I ran `npm run build:css` and I think that fixed both of the problems I'd identified! The line is gone, and the column width for tags is wider now. 

![image](https://github.com/user-attachments/assets/4b93b116-0880-4060-83d4-ad862c8a63bb)

----

It looks like you also changed the column width for comments, which looks a little strange in this case with only brief comments. 

![image](https://github.com/user-attachments/assets/b6125770-3882-449f-a964-b4f57039211a)

But it will look fine in most normal uses!

![image](https://github.com/user-attachments/assets/7497f8ac-3c2d-45da-ac20-d3385dd8c9d1)

rajaro commented:
@demiankatz Well that explains it! I did not add the compiled.css, which I now did. Also, regarding Mink test case, I'm having some issues with my local installation setup with regards to mink tests. If you happen to have the time to write up a mink test for this it would be great, but I will get back to this next week.

demiankatz commented:
@rajaro, I'm catching up after being out of office for a few days, so I probably won't get to add a Mink test here right away. If I do find time to start the work, I will comment here when I begin. If you get there first, please let me know when you start -- I just want to be sure we don't both work on it at the same time. :-)

rajaro commented:
@demiankatz I was just about comment on this: got my setup working will work on the mink tests.

demiankatz commented:
Excellent, thanks, @rajaro! Let me know if I can be of any further assistance.

rajaro committed:
Add mink test case

demiankatz commented on module/VuFind/tests/integration-tests/src/VuFindTest/Mink/RecordActionsTest.php:
```diff
@@ -946,6 +946,82 @@ function () use ($page, $checked) {
+        $this->clickCss($page, '.checkbox input[name="selectAll"]');
+        $this->clickCss($page, 'button#cancelSelected');
+        $this->clickCss($page, 'a#confirm_cancel_selected_yes');
+        $this->unfindCss($page, 'usercontent-table');
```
Should this be:

```suggestion
        $this->unfindCss($page, '.usercontent-table');
```

?

demiankatz commented on module/VuFind/tests/integration-tests/src/VuFindTest/Mink/RecordActionsTest.php:
```diff
@@ -946,6 +946,82 @@ function () use ($page, $checked) {
+        $this->clickCss($page, '.checkbox input[name="selectAll"]');
+        $this->clickCss($page, 'button#cancelSelected');
+        $this->clickCss($page, 'a#confirm_cancel_selected_yes');
+        $this->unfindCss($page, 'usercontent-table');
```
Same here:

```suggestion
        $this->unfindCss($page, '.usercontent-table');
```

demiankatz commented on module/VuFind/tests/integration-tests/src/VuFindTest/Mink/RecordActionsTest.php:
```diff
@@ -946,6 +946,82 @@ function () use ($page, $checked) {
+        $this->clickCss($page, '.checkbox input[name="selectAll"]');
+        $this->clickCss($page, 'button#cancelSelected');
+        $this->clickCss($page, 'a#confirm_cancel_selected_yes');
+        $this->unfindCss($page, 'usercontent-table');
```
And here:

```suggestion
        $this->unfindCss($page, '.usercontent-table');
```

demiankatz commented on module/VuFind/tests/integration-tests/src/VuFindTest/Mink/RecordActionsTest.php:
```diff
@@ -946,6 +946,82 @@ function () use ($page, $checked) {
+    public function testUserContentDeletion()
+    {
+        $page = $this->gotoRecord();
+        $this->changeConfigs(
```
It might be better to change the configuration before going to the record page, to ensure that the configuration is constant for the entire test.

demiankatz commented on module/VuFind/tests/integration-tests/src/VuFindTest/Mink/RecordActionsTest.php:
```diff
@@ -946,6 +946,82 @@ function () use ($page, $checked) {
+        $this->fillInLoginForm($page, 'username1', 'test');
+        $this->submitLoginForm($page);
+
+        $page = $this->gotoRecord();
```
Do you need this extra "goToRecord" call? Might it only have been needed in the past because of the configuration order issue I mentioned above?

demiankatz commented on module/VuFind/tests/integration-tests/src/VuFindTest/Mink/RecordActionsTest.php:
```diff
@@ -946,6 +946,82 @@ function () use ($page, $checked) {
+
+        $this->clickCss($page, 'li#user-content-ratings a.nav-link');
+        $this->waitForPageLoad($page);
+        $this->findCss($page, '.usercontent-table');
```
Is it worth adding an assertion here to check for the rated title, or the actual rating value, to be extra sure things are as expected?

demiankatz commented on module/VuFind/tests/integration-tests/src/VuFindTest/Mink/RecordActionsTest.php:
```diff
@@ -946,6 +946,82 @@ function () use ($page, $checked) {
+
+        $this->clickCss($page, 'li#user-content-tag a.nav-link');
+        $this->waitForPageLoad($page);
+        $this->findCss($page, '.usercontent-table');
```
Should we also search for `testtag` here?

demiankatz reviewed (changes requested):
Thanks for the test, @rajaro -- looks like a good start, but I have a few suggestions for improvements. Let me know if you need help with any of the details. Once this is all addressed, I'll try running the test locally to be sure it passes in my environment as well.

rajaro committed:
Improve test case

rajaro committed:
fix merge conflicts

demiankatz commented:
Thanks, @rajaro, your test improvements look good to me!

Now there are just a couple of questions:

1.) I see that @sturkel89 has an outstanding "request changes" review. Susan, were you just waiting for me to finish reviewing the test before proceeding, or are there still some outstanding issues on your end?

2.) Since the Doctrine conversion work in #2233 is finally nearing completion, I'm wondering if it would be easiest to wait until that is done before merging this, so we can reconcile the Laminas to Doctrine conversion in this branch, instead of adding more work to the in-progress work on #2233. Is there any urgency on your end to get this merged quickly, @rajaro, or would you be okay with putting this on hold until Doctrine is done?

demiankatz reviewed (changes requested):
@rajaro, I tried to actually run the test in my environment, and I'm getting this failure:

```
1) VuFindTest\Mink\RecordActionsTest::testUserContentDeletion
Exception: Failed to call getText on '.alert-success' after 6 attempts.

/usr/local/vufind/module/VuFind/src/VuFindTest/Integration/MinkTestCase.php:828
/usr/local/vufind/module/VuFind/src/VuFindTest/Integration/MinkTestCase.php:747
/usr/local/vufind/module/VuFind/tests/integration-tests/src/VuFindTest/Mink/RecordActionsTest.php:992
```

Are you seeing the same thing on your end, or do you need me to investigate further on mine?

sturkel89 reviewed (approved):
I've done a quick re-test of this PR. Everything loaded fine and the interface looks good. Long comments get the expander, and short comments display correctly. All columns are wide enough. No stray lines appear in the interface. I think it's looking great and ready to go, once the tests are fixed!

rajaro committed:
Slight mink test adjustment

rajaro commented:
@demiankatz I'm not seeing that on my end. I made a slight change to the test, hopefully it should now pass on your end. If it doesn't, I will look more into my testing setup. 

Also, if possible, it would be great if we can get this merged soon

demiankatz commented:
@rajaro, the test is still failing for me (though in a slightly different way):

```
1) VuFindTest\Mink\RecordActionsTest::testUserContentDeletion
Element not found: .modal form div.star-rating label index 5
Failed asserting that null is of type object.

/usr/local/vufind/module/VuFind/src/VuFindTest/Integration/MinkTestCase.php:561
/usr/local/vufind/module/VuFind/src/VuFindTest/Integration/MinkTestCase.php:655
/usr/local/vufind/module/VuFind/tests/integration-tests/src/VuFindTest/Mink/RecordActionsTest.php:990
```

Here is a screen shot of the state of the test when it failed:

![testUserContentDeletion-12708302217627](https://github.com/user-attachments/assets/8776b833-2920-460b-b88f-1b8f10f06ba8)

It looks like the problem is that the rating control failed to open in a modal for some reason.

Is that any help? Please let me know if you'd like me to try manually stepping through the test to see if I can work out further details.

rajaro committed:
Fix mink test

rajaro commented:
@demiankatz I now saw the same issue where the rating modal failed to open, but I managed to fix it by adding a rating before tags or comments. Not really sure what causes the issue

demiankatz commented:
Thanks, @rajaro, the test is passing now.

I did a bit of experimentation, and it looks to me like this has revealed a previously-unnoticed bug: if you add a comment, it breaks the ratings pop-up for some reason. I've opened [VUFIND-1773](https://openlibraryfoundation.atlassian.net/browse/VUFIND-1773) to track the issue. I don't think this breaks functionality (it's just an inconsistent behavior), and I don't use ratings at my institution, so this isn't going to be a high priority for me to fix... but I thought it was worth tracking in case others are more concerned about it!

I think that means that this PR is now ready to merge. However, I would like to make sure that it is fully Doctrine-ready before making the merge. Do you have time to create a branch against [the Doctrine work in progress](https://github.com/demiankatz/vufind/tree/doctrine) that incorporates this work? That would allow me to merge this sooner, and then immediately reincorporate it into the doctrine branch. If you don't have time for that, I'm willing to help with the work -- but this will have to stay on hold for a little longer as I finish up other Doctrine-related preparations.

demiankatz commented on module/VuFind/tests/integration-tests/src/VuFindTest/Mink/RecordActionsTest.php:
```diff
@@ -946,6 +946,84 @@ function () use ($page, $checked) {
+        $this->waitForPageLoad($page);
+        $this->clickCss($page, '#loginOptions a');
+        $this->findCss($page, $this->openModalUsernameFieldSelector);
+        $this->makeAccount($page, 'username5');
```
One more small issue: you need to clean up `username5` in `tearDownAfterClass` below, or else subsequent tests will fail.

demiankatz reviewed (changes requested):
One more minor issue that I just discovered when running the full test suite...

rajaro commented:
@demiankatz Sure, I'll make a PR with the necessary changes for Doctrine. Didn't have enough time to finish it but I'll get back to it next week.

rajaro committed:
mink test cleanup

demiankatz committed:
Merge branch 'dev' into user-comments-ratings

demiankatz reviewed (commented):
Thanks, @rajaro, this is looking good to me, and I think we should be able to merge it as-is. I also appreciate the work you've invested in creating the Doctrine-friendly PR; that will also be extremely helpful!

Since this PR adds new language strings and we're currently in the middle of preparing translations for the 10.2 release, I will have to hold off on merging this until that work is completed, to prevent things from getting into a confusing state on the Lokalise platform. That means I should be able to merge this the week of June 23 at the latest.

In the meantime, one minor question/suggestion: should we display a message in the tabs that have no data, perhaps reusing the "No tags" message and creating new "No comments" and "No ratings" messages, so it's clear that no data is available? I don't think this is strictly necessary, but I wonder if it would make the interface feel slightly more polished.

rajaro committed:
add no tags/comments/ratings text

demiankatz committed:
Merge branch 'dev' into user-comments-ratings

demiankatz reviewed (approved):
Thanks again, @rajaro -- this is working well, and now that release 10.2 is done, I can merge it.
